drivers/crypto/crypto_si32.c

```diff
@@ -0,0 +1,1157 @@
+
+	/* To use the AES0 module, firmware must first clear the RESET bit before initializing the
+	 * registers.
+	 */
+	SI32_AES_A_reset_module(SI32_AES_0);
+
+	__ASSERT(SI32_AES_0->CONTROL.RESET == 0, "Reset done");
+
+	/* 12.3. Interrupts: The completion interrupt should only be used in conjunction
+	 * with software mode (SWMDEN bit is set to 1) and not with DMA operations, where the DMA
+	 * completion interrupt should be used.
+	 */
+	SI32_AES_A_disable_operation_complete_interrupt(SI32_AES_0); /* default */
+
+	/* 12.3. Interrupts: The error interrupt should always be enabled (ERRIEN = 1), even when
+	 * using the DMA with the AES module.
+	 */
+	SI32_AES_A_enable_error_interrupt(SI32_AES_0);
+
+	/* Install error handler */
+	IRQ_CONNECT(DT_INST_IRQN(0), DT_INST_IRQ(0, priority), crypto_si32_irq_error_handler,
+		    DEVICE_DT_INST_GET(0), 0);
+	irq_enable(DT_INST_IRQN(0));
+
+	/* Halt AES0 module on debug breakpoint */
+	SI32_AES_A_enable_stall_in_debug_mode(SI32_AES_0);
+
+	/* For peripheral transfers, firmware should configure the peripheral for the DMA transfer
+	 * and set the device’s DMA crossbar (DMAXBAR) to map a DMA channel to the peripheral.
+	 */
+	SI32_DMAXBAR_A_select_channel_peripheral(SI32_DMAXBAR_0, SI32_DMAXBAR_CHAN5_AES0_TX);
+	SI32_DMAXBAR_A_select_channel_peripheral(SI32_DMAXBAR_0, SI32_DMAXBAR_CHAN6_AES0_RX);
+	SI32_DMAXBAR_A_select_channel_peripheral(SI32_DMAXBAR_0, SI32_DMAXBAR_CHAN7_AES0_XOR);
+
+	return 0;
+}
+
+static int crypto_si32_aes_set_key(const uint8_t *key, uint8_t key_len)
+{
+	switch (key_len) {
+	case 32:
+		SI32_AES_0->HWKEY7.U32 = *((uint32_t *)key + 7);
+		SI32_AES_0->HWKEY6.U32 = *((uint32_t *)key + 6);
+	case 24:
+		SI32_AES_0->HWKEY5.U32 = *((uint32_t *)key + 5);
+		SI32_AES_0->HWKEY4.U32 = *((uint32_t *)key + 4);
+	case 16:
+		SI32_AES_0->HWKEY3.U32 = *((uint32_t *)key + 3);
+		SI32_AES_0->HWKEY2.U32 = *((uint32_t *)key + 2);
+		SI32_AES_0->HWKEY1.U32 = *((uint32_t *)key + 1);
+		SI32_AES_0->HWKEY0.U32 = *((uint32_t *)key);
+		break;
+	default:
+		LOG_ERR("Invalid key len: %" PRIu16, key_len);
+		return -EINVAL;
+	}
+
+	return 0;
+}
+
+static int crypto_si32_aes_calc_decryption_key(const struct cipher_ctx *ctx,
+					       uint8_t *decryption_key)
+{
+	int ret;
+
+	ret = crypto_si32_aes_set_key(ctx->key.bit_stream, ctx->keylen);
+	if (ret) {
+		return ret;
+	}
+
+	LOG_INF("Generating decryption key");
+	/* TODO: How much of this can be removed? */
+	SI32_AES_A_write_xfrsize(SI32_AES_0, 0);
+	SI32_AES_A_enable_error_interrupt(SI32_AES_0);
+	SI32_AES_A_exit_cipher_block_chaining_mode(SI32_AES_0);
+	SI32_AES_A_exit_counter_mode(SI32_AES_0);
+	SI32_AES_A_exit_bypass_hardware_mode(SI32_AES_0);
+	SI32_AES_A_select_xor_path_none(SI32_AES_0);
+	SI32_AES_A_select_software_mode(SI32_AES_0);
+	SI32_AES_A_select_encryption_mode(SI32_AES_0);
+	SI32_AES_A_enable_key_capture(SI32_AES_0);
+
+	for (int_fast8_t i = 0; i < 4; i++) {
```

Comment:
Unless you are able to justify it, I suggest to not clutter with types like `int_fast8_t`. An `int` is probably right here.

Reply:
Done, but went for unsigned int.